[router] Impl radix tree and set up CI

ByronHsu opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

1. implement a radix tree for later use
2. set up CI for rust

## Modifications

<!-- Describe the changes made in this PR. -->

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

ByronHsu committed:
add rust ci and radix tree impl

wip

radix tree

wip

test ci

add rust ci

add rust ci

add rust ci

add rust ci

add rust ci

add rust ci

add rust ci

add comment

resolve

merrymercy commented on .github/workflows/pr-test-rust.yml:
```diff
@@ -0,0 +1,35 @@
+  workflow_dispatch:
+
+concurrency:
+  group: pr-test-${{ github.ref }}
```
Use a different name to resolve conflict. Otherwise, it will cancel other workflows.

ByronHsu replied:
fixed

ByronHsu replied:
@merrymercy do you know what is the pending "finish" in the ci?

Ying1123 commented on rust/src/tree.rs:
```diff
@@ -0,0 +1,214 @@
+#[derive(Clone)]
+pub struct Node {
+    pub children: Vec<Node>,
```
Use dict of [first char] -> Nod, rather than vector.

ByronHsu replied:
Made the change. great suggestion

Ying1123 commented on rust/src/tree.rs:
It's fine as a vanilla version for now. In the future, we'll need to support multiple processes/threads to access the tree. So race condition should be avoided (e.x. using Arc and RwLock).

ByronHsu replied:
Will do in the follow-up PRs!

Ying1123 commented on rust/src/tree.rs:
```diff
@@ -0,0 +1,214 @@
+
+    }
+
+    pub fn delete(&mut self, input_ids: &[usize]) {
```
In the future PRs, we might want to support delete with node contractions.

ByronHsu replied:
makes sense

ByronHsu committed:
remove clone

ByronHsu committed:
add fmt

ByronHsu committed:
add fmt ci

ByronHsu commented:
@Ying1123 I have 
1. Fixed the `clone()` redundancy by `mem::take`
2. Replaced vector with hashmap
3. Added fmt check in ci

Thanks for the great feedback

ByronHsu committed:
update md

Ying1123 commented on rust/src/tree.rs:
```diff
@@ -0,0 +1,215 @@
+        while curr_idx < input_ids_len {
+            let mut match_info = None;
+
+            for key in curr.children.keys() {
```
Since you have changed the children to be a hashmap. The for loop is not needed here.

Ying1123 commented on rust/src/tree.rs:
```diff
@@ -0,0 +1,215 @@
+            let mut child_key = None;
+            let mut prefix_len = 0;
+
+            for key in curr.children.keys() {
```
Same here

ByronHsu committed:
remove for

ByronHsu committed:
optimize

ByronHsu committed:
Merge branch 'main' into byhsu/rust-approx-tree

ByronHsu committed:
Merge branch 'main' into byhsu/rust-approx-tree

ByronHsu committed:
Merge branch 'main' into byhsu/rust-approx-tree

merrymercy commented on .github/workflows/pr-test-rust.yml:
```diff
@@ -0,0 +1,39 @@
+name: PR Test
```
```suggestion
name: PR Test (Rust)
```

ByronHsu committed:
Update .github/workflows/pr-test-rust.yml

Co-authored-by: Lianmin Zheng <lianminzheng@gmail.com>
